aya/src/maps/xdp/cpu_map.rs

```diff
@@ -0,0 +1,151 @@
+
+        let _fd = data.fd_or_err()?;
+
+        Ok(CpuMap { inner: map })
+    }
+
+    /// Returns the number of elements in the array.
+    ///
+    /// This corresponds to the value of `bpf_map_def::max_entries` on the eBPF side.
+    pub fn len(&self) -> u32 {
+        self.inner.as_ref().obj.max_entries()
+    }
+
+    /// Returns the value stored at the given index.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`MapError::OutOfBounds`] if `index` is out of bounds, [`MapError::SyscallError`]
+    /// if `bpf_map_lookup_elem` fails.
+    pub fn get(&self, index: u32, flags: u64) -> Result<CpuMapValue, MapError> {
+        let data = self.inner.as_ref();
+        check_bounds(data, index)?;
+        let fd = data.fd_or_err()?;
+
+        let value = bpf_map_lookup_elem(fd, &index, flags).map_err(|(_, io_error)| {
+            MapError::SyscallError {
+                call: "bpf_map_lookup_elem".to_owned(),
+                io_error,
+            }
+        })?;
+        let value: bpf_cpumap_val = value.ok_or(MapError::KeyNotFound)?;
+        Ok(value.into())
+    }
+
+    /// An iterator over the elements of the map. The iterator item type is `Result<u32,
+    /// MapError>`.
+    pub fn iter(&self) -> impl Iterator<Item = Result<CpuMapValue, MapError>> + '_ {
+        (0..self.len()).map(move |i| self.get(i, 0))
+    }
+}
+
+impl<T: AsMut<MapData>> CpuMap<T> {
+    /// Sets the value of the element at the given index.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`MapError::OutOfBounds`] if `index` is out of bounds, [`MapError::SyscallError`]
+    /// if `bpf_map_update_elem` fails.
+    pub fn set(
+        &mut self,
+        index: u32,
+        value: u32,
+        program: Option<impl AsRawFd>,
+        flags: u64,
+    ) -> Result<(), MapError> {
+        let data = self.inner.as_mut();
+        check_bounds(data, index)?;
+        let fd = data.fd_or_err()?;
+
+        let value = bpf_cpumap_val {
+            qsize: value,
+            bpf_prog: bpf_cpumap_val__bindgen_ty_1 {
+                fd: program.map(|prog| prog.as_raw_fd()).unwrap_or_default(),
+            },
+        };
+        bpf_map_update_elem(fd, Some(&index), &value, flags).map_err(|(_, io_error)| {
+            MapError::SyscallError {
+                call: "bpf_map_update_elem".to_owned(),
+                io_error,
+            }
+        })?;
+        Ok(())
+    }
+}
+
+impl<T: AsRef<MapData>> IterableMap<u32, CpuMapValue> for CpuMap<T> {
+    fn map(&self) -> &MapData {
+        self.inner.as_ref()
+    }
+
+    fn get(&self, key: &u32) -> Result<CpuMapValue, MapError> {
+        self.get(*key, 0)
+    }
+}
+
+unsafe impl Pod for bpf_cpumap_val {}
+
+#[derive(Clone, Copy, Debug)]
+pub struct CpuMapValue {
+    pub qsize: u32,
+    pub prog_id: u32,
+}
+
+impl From<bpf_cpumap_val> for CpuMapValue {
+    fn from(value: bpf_cpumap_val) -> Self {
+        // SAFETY: map writes use fd, map reads use id.
```

Comment:
The fact that this requires a SAFETY comment is a good indication that we should
probably not use From to convert, as it seems like a possible future footgun